druid/src/widget/split.rs

```diff
@@ -0,0 +1,327 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget which splits an area in two, with a set ratio.
+
+use crate::kurbo::{Line, Point, Rect, Size};
+use crate::widget::flex::Axis;
+use crate::{
+    theme, BaseState, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, PaintCtx,
+    RenderContext, UpdateCtx, Widget, WidgetPod,
+};
+
+///A container containing two other widgets, splitting the area either horizontally or vertically.
+pub struct Split<T: Data> {
+    split_point: f64,
+    split_direction: Axis,
+    splitter_size: f64, // Should this be a u32?
+    draggable: bool,
+    child1: WidgetPod<T, Box<dyn Widget<T>>>,
+    child2: WidgetPod<T, Box<dyn Widget<T>>>,
+}
+
+/// A builder for a widget which divides space horizontally
+/// The splitter bar will run top-to-bottom
+///
+/// The actual widget it implemented by [`Split`], but this builder
+/// is provided for convenience.
+///
+/// [`Split`]: struct.Split.html
+pub struct HorizontalSplit;
+impl HorizontalSplit {
+    /// Create a new split panel which splits the horizontal area,
+    /// leaving the full height available to both children.
+    ///
+    /// Note that `split_point` must be between 0 and 1  
+    pub fn new<T: Data>(
+        split_point: f64,
+        splitter_size: f64,
+        draggable: bool,
+        child1: impl Widget<T> + 'static,
+        child2: impl Widget<T> + 'static,
+    ) -> Split<T> {
+        Split::new(
+            Axis::Horizontal,
+            split_point,
+            splitter_size,
+            draggable,
+            child1,
+            child2,
+        )
+    }
+}
+/// A builder for a widget which divides space vertically
+/// The splitter bar will run left-to-right
+///
+/// The actual widget it implemented by [`Split`], but this builder
+/// is provided for convenience.
+///
+/// [`Split`]: struct.Split.html
+pub struct VerticalSplit;
+impl VerticalSplit {
+    /// Create a new split panel which splits the vertical area,
+    /// leaving the full width available to both children.
+    ///
+    /// Note that `split_point` must be between 0 and 1  
+    pub fn new<T: Data>(
+        split_point: f64,
+        splitter_size: f64,
+        draggable: bool,
+        child1: impl Widget<T> + 'static,
+        child2: impl Widget<T> + 'static,
+    ) -> Split<T> {
+        Split::new(
+            Axis::Vertical,
+            split_point,
+            splitter_size,
+            draggable,
+            child1,
+            child2,
+        )
+    }
+}
+
+impl<T: Data> Split<T> {
+    ///Create a new split panel.
+    ///Note that split_point must be between 0 and 1.
+    fn new(
```

Comment:
Okay, my only main nit now is around constructors, where I have a few thoughts.

- Per the discussion in zulip, I think that `Split::vertical` and `Split::horizontal` constructors make sense. If we want to keep the `VerticalSplit` and `HorizontalSplit` structs, their `new` fns can just call the corresponding constructors on `Split`.
- With widgets that can take a number of arguments at init time, I think a builder pattern makes sense. For an example of this, see [`SizedBox`](https://github.com/xi-editor/druid/blob/master/druid/src/widget/sized_box.rs#L59). In this case, I think that `draggable` and `splitter_size` probably make sense as optional builder-style methods.
- `splitter_size` is an interesting one; this might actually be something that we want to put in the `env`? Or we might want to put the default value in the `env`? This comes down to whether or not we expect to be tuning this per-platform. I'm okay with punting on this for now.

Reply:
I had punted on adjusting the types-versus-multiple-constructors based on where we'd landed in zulip - I thought you and I agreed, but the consensus was 🤷. I still do like the constructors method, so I can refactor that way.

I can move to builder-style functions here, if that's preferable. Do we want to allow for not setting children? My reasoning for requiring them in `::new()` was that you can always put some blank widget as a child, if you want a blank space, but it ups the complexity significantly to handle the `None` case in each of our methods.

I definitely have no objection to `splitter_size` and `draggable` as builder methods.

I think some more discussion on if/how we scope `Env` and what goes in it is a good idea. I can see something like a `size::PADDING` (default_padding? border_width?), and the splitter would presumably be two of those wide/tall.

Reply:
I think we definitely always want both children to be specified. It's possible that we don't always want `splitter_position`, (and it's possible we want to allow specifying it in pixels) but I'd been assuming we would always take it, as well; and then we would just have the two 'more optional' arguments be builder methods.

Reply:
Refactored for `::vertical` and `::horizontal` constructors.
Added builder-style methods for `split_point`, `splitter_width`, `draggable`, with defaults of 0.5, 10.0, and false, respectively. Punted on setting the split point in pixels from an edge - unless we're in a `SizedBox`, we don't know our size at construction time (and even then, we could end up in a smaller size)